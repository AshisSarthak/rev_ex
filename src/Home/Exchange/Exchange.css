.exchange-from-container {
  display: flex;
  width: 80%;
  justify-content: space-around;
  align-items: center;
  margin-top: -10px;
  margin-left: 10%;
  height: 200px;
  border: 1px solid #10cbf7;
  border-radius: 10px;
}

.exchange-from-select {
  font-size: 33px;
  border: 1px solid transparent;
  border-radius: 10px;
  padding: 10px;
  outline: none;
}
.exchange-from-input {
  font-size: 33px;
  padding: 10px;
  border-radius: 4px;
  border: 1px solid transparent;
  outline: none;
  margin-left: 10px;
  width: 220px;
}

.exchange-rate {
  position: relative;
  bottom: -10px;
  background-color: #10cbf7;
  width: 20%;
  left: calc((100% - 20%) / 2);
  padding: 5px;
  text-align: center;
  border-radius: 4px;
  font-size: 20px;
}

.exchange-to-label {
  font-size: 4em;
  font-weight: 100;
  color: white;
}

.convert-button {
  float: right;
  padding: 10px;
  border: 1px solid transparent;
  border-radius: 4px;
  cursor: pointer;
  margin: 20px 200px 0px 0px;
  font-size: 20px;
  background-color: #10cbf7;
}
