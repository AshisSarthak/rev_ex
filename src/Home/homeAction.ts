import { EXCHANGE_API } from "./homeConstants";
import { RawFX } from "./homeTypes";

export const fetchCurrentFXRateAction = (rawFxRates: RawFX) => ({
  type: "FETCH_FX_RATE",
  payload: rawFxRates,
});

export const fetchCurrentFXRate = (baseCurrency: string = "USD") => async (
  dispatch: any
): Promise<void> => {
  const fxRates: RawFX = await fetch(
    `${EXCHANGE_API}${baseCurrency}`
  ).then((resp) => resp.json());
  console.log(fxRates);
  dispatch(fetchCurrentFXRateAction(fxRates));
};
