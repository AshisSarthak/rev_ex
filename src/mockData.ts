export const demoPacketData = {
  balance: 100,
  transactions: [
    {
      id: 1,
      timestamp: 1000230,
      balance: 50,
    },
    {
      id: 2,
      timestamp: 1000230,
      balance: 80,
    },
    {
      id: 3,
      timestamp: 1000230,
      balance: 90,
    },
  ],
};

export enum CURRENCY_SYMBOL {
  USD = "$",
  EUR = "€",
  GBP = "£",
}

export enum CURRENCYS {
  USD = "USD",
  EUR = "EUR",
  GBP = "GBP",
}
