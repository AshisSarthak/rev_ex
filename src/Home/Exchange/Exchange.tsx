import React from "react";
import { connect } from "react-redux";
import { fetchCurrentFXRate, convertToPocket } from "../homeAction";
import { CURRENCYS } from "../../mockData";
import SelectEl from "../Common/Select";
import InputEl from "../Common/Input";
import "./Exchange.css";
import { getCurrencySymbol } from "../Pockets/Pockets";

const defaultOptionList = [CURRENCYS.USD, CURRENCYS.EUR, CURRENCYS.GBP];
const validInputReg = new RegExp(/^[+-]?((\d+(\.\d{1,2})?)|(\.\d+))$/);

export const getToCurrencyList = (targetCur: CURRENCYS) =>
  defaultOptionList.filter((val: CURRENCYS) => val !== targetCur);

export type ExchangeState = {
  fromCurrency: CURRENCYS;
  fromAmount: number | undefined;
  toCurrencyList: Array<CURRENCYS>;
  toCurrency: CURRENCYS;
  toAmount: number | undefined;
};

export type Props = {
  fetchCurrentFXRate: Function;
  convertToPocket: Function;
  fxRate: any;
};

export class Exchange extends React.Component<Props, ExchangeState> {
  state = {
    fromCurrency: CURRENCYS.USD,
    toCurrencyList: getToCurrencyList(CURRENCYS.USD),
    toCurrency: getToCurrencyList(CURRENCYS.USD)[0],
    toAmount: undefined,
    fromAmount: undefined,
  };

  componentDidMount() {
    this.props.fetchCurrentFXRate(this.state.fromCurrency);
  }

  selectFromChangeHandler = async (value: CURRENCYS) => {
    await this.props.fetchCurrentFXRate(value);
    this.setState({
      fromCurrency: value,
      toCurrencyList: getToCurrencyList(value),
      toCurrency: getToCurrencyList(value)[0],
      toAmount:
        this.state.fromAmount! *
        this.getExchangeVal(getToCurrencyList(value)[0]),
    });
  };

  selectToChangeHandler = (value: any) => {
    this.setState({
      toCurrency: value,
      toAmount: this.state.fromAmount! * this.getExchangeVal(value),
    });
  };

  selectFromHandler = (value: any) => {
    this.setState({
      fromAmount: value,
      toAmount: value * this.getExchangeVal(),
    });
  };

  getExchangeVal = (targetCurrency: CURRENCYS | undefined = undefined) => {
    if (this.props.fxRate && this.props.fxRate.rates) {
      return targetCurrency
        ? this.props.fxRate.rates[targetCurrency!]
        : this.props.fxRate.rates[this.state.toCurrency];
    }
  };

  handleConvertPocket = () => {
    const { fromAmount, fromCurrency, toAmount, toCurrency } = this.state;
    this.props.convertToPocket({
      fromAmount,
      fromCurrency,
      toAmount,
      toCurrency,
    });
  };

  render() {
    const exchangeRateString =
      this.getExchangeVal() &&
      `1${getCurrencySymbol(
        this.state.fromCurrency
      )} = ${this.getExchangeVal()}${getCurrencySymbol(this.state.toCurrency)}`;

    return (
      <section className="exchange-container">
        <div className="exchange-rate">{exchangeRateString}</div>
        <div className="exchange-from-container">
          <div className="exchange-box">
            <SelectEl
              onChange={this.selectFromChangeHandler}
              optionLists={defaultOptionList}
              currentVal={this.state.fromCurrency}
            />
            <InputEl
              onChange={this.selectFromHandler}
              value={this.state.fromAmount}
            />
          </div>
          <div className="exchange-to-label"> to </div>
          <div className="exchange-box">
            <SelectEl
              onChange={this.selectToChangeHandler}
              optionLists={this.state.toCurrencyList}
              currentVal={this.state.toCurrency}
            />
            <InputEl value={this.state.toAmount} disabled />
          </div>
        </div>

        <button
          className="convert-button"
          disabled={!validInputReg.test(this.state.fromAmount!)}
          onClick={this.handleConvertPocket}
        >
          Convert to Pocket
        </button>
      </section>
    );
  }
}

const mapStateToProps = (state: any) => ({
  fxRate: state.fxRates.fxRates,
});

const mapDispatchToProps = {
  fetchCurrentFXRate,
  convertToPocket,
};

export default connect(mapStateToProps, mapDispatchToProps)(Exchange);
