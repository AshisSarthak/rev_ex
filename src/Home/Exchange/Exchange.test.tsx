import React from "react";
import fetchMock from "jest-fetch-mock";
import { shallow } from "enzyme";
import { Exchange } from "./Exchange";
import { mockResponse } from "../homeAction.test";
import SelectEl from "../Common/Select";
import { CURRENCYS } from "../../mockData";
import InputEl from "../Common/Input";

describe("Exchange Section", () => {
  const currencyChangeHandler = jest.fn();
  const covertActionHandler = jest.fn();
  const rateMap: Map<string, number> = new Map();
  rateMap.set("USD", 10);
  rateMap.set("GBP", 10);
  rateMap.set("EUR", 10);

  const staticFXRate = {
    rates: rateMap,
    base: "USD",
    date: "122333434",
  };
  it("Loads the default Exchange section", () => {
    const wrapper = shallow(
      <Exchange
        fxRate={staticFXRate}
        fetchCurrentFXRate={currencyChangeHandler}
        convertToPocket={covertActionHandler}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("Change in From Currency dropdown", async () => {
    const response = { ...mockResponse };
    fetchMock.once(JSON.stringify(response));
    const wrapper = shallow(
      <Exchange
        fxRate={staticFXRate}
        fetchCurrentFXRate={currencyChangeHandler}
        convertToPocket={covertActionHandler}
      />
    );

    const selectEls = wrapper.find(SelectEl);
    expect(selectEls).toHaveLength(2);
    expect(wrapper.state("toCurrencyList")).toMatchObject([
      CURRENCYS.EUR,
      CURRENCYS.GBP,
    ]);
    expect(wrapper.state("toCurrency")).toBe(CURRENCYS.EUR);
    expect(wrapper.state("toAmount")).toBeUndefined();
    expect(wrapper.state("fromAmount")).toBeUndefined();

    const fromSelect = selectEls.at(0);
    fromSelect.simulate("change", CURRENCYS.EUR);
    expect(currencyChangeHandler).toBeCalledWith(CURRENCYS.EUR);

    await wrapper.update();
    expect(wrapper.state("toCurrency")).toBe(CURRENCYS.USD);
    expect(wrapper.state("toAmount")).toBeNaN();
    expect(wrapper.state("fromAmount")).toBeUndefined();
    expect(wrapper.state("toCurrencyList")).toMatchObject([
      CURRENCYS.USD,
      CURRENCYS.GBP,
    ]);

    wrapper.setState({
      fromAmount: 10,
    });
    await wrapper.update();
    expect(wrapper.state("toAmount")).toBe(staticFXRate[CURRENCYS.EUR] * 10);
  });

  it("Change in the From Amount text box", async () => {
    const wrapper = shallow(
      <Exchange
        fxRate={staticFXRate}
        fetchCurrentFXRate={currencyChangeHandler}
        convertToPocket={covertActionHandler}
      />
    );
    const inputEls = wrapper.find(InputEl);
    expect(inputEls).toHaveLength(2);

    const fromInput = inputEls.at(0);
    fromInput.simulate("change", 5);
    await wrapper.update();
    expect(wrapper.state("toAmount")).toBe(
      staticFXRate.rates[wrapper.state("toCurrency")] * 5
    );

    fromInput.simulate("change", "asdd");
    await wrapper.update();
    expect(wrapper.state("toAmount")).toBeNaN();
  });

  it("Change in the To Currency dropdown", async () => {
    const wrapper = shallow(
      <Exchange
        fxRate={staticFXRate}
        fetchCurrentFXRate={currencyChangeHandler}
        convertToPocket={covertActionHandler}
      />
    );
    const selectEls = wrapper.find(SelectEl);

    const toCurr = selectEls.at(1);
    toCurr.simulate("change", CURRENCYS.GBP);

    await wrapper.update();
    expect(wrapper.state("toCurrency")).toBe(CURRENCYS.GBP);
  });

  it("Handle click of convert button", async () => {
    const wrapper = shallow(
      <Exchange
        fxRate={staticFXRate}
        fetchCurrentFXRate={currencyChangeHandler}
        convertToPocket={covertActionHandler}
      />
    );
    const convertBtn = wrapper.find("button").at(0);
    convertBtn.simulate("click");
    expect(covertActionHandler).toBeCalledWith({
      fromAmount: wrapper.state("fromAmount"),
      fromCurrency: wrapper.state("fromCurrency"),
      toAmount: wrapper.state("toAmount"),
      toCurrency: wrapper.state("toCurrency"),
    });
  });
});
