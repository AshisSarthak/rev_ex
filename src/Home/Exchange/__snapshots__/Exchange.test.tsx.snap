// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exchange Section Loads the default Exchange section 1`] = `
<section
  className="exchange-container"
>
  <div
    className="exchange-rate"
  />
  <div
    className="exchange-from-container"
  >
    <div
      className="exchange-box"
    >
      <SelectEl
        currentVal="USD"
        onChange={[Function]}
        optionLists={
          Array [
            "USD",
            "EUR",
            "GBP",
          ]
        }
      />
      <InputEl
        onChange={[Function]}
      />
    </div>
    <div
      className="exchange-to-label"
    >
       to 
    </div>
    <div
      className="exchange-box"
    >
      <SelectEl
        currentVal="EUR"
        onChange={[Function]}
        optionLists={
          Array [
            "EUR",
            "GBP",
          ]
        }
      />
      <InputEl
        disabled={true}
      />
    </div>
  </div>
  <button
    className="convert-button"
    disabled={true}
    onClick={[Function]}
  >
    Convert to Pocket
  </button>
</section>
`;
